compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -1059,6 +1056,27 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
     /// Assembles the trait which are built-in to the language itself:
     /// `Copy`, `Clone` and `Sized`.
     #[instrument(level = "debug", skip(self, candidates))]
+    fn assemble_builtin_sized_candidate(
+        &mut self,
+        obligation: &PolyTraitObligation<'tcx>,
+        candidates: &mut SelectionCandidateSet<'tcx>,
+    ) {
+        match self.sized_conditions(obligation) {
+            BuiltinImplConditions::Where(nested) => {
+                candidates
+                    .vec
+                    .push(SizedCandidate { has_nested: !nested.skip_binder().is_empty() });
```

Comment:
I have convinced myself this is a non-issue, but just to get some more grey matter on it:

Should we ever do the inverse for unsized types, do you forsee any weird interactions.

```rust
#![feature(trivial_bounds)]
trait Trait: Sized {
    const CONST: () = ();
}

impl<T: Sized> Trait for T {}

fn foo()
where
    str: Sized,
{
    <str as Trait>::CONST;
}
```

This would change behaviour, which is fine, but there may be other cases I didn't consider

Reply:
what's the inverse here? and what's your expected behavior for that test
```

Reply:
If we had a builtin `!Sized` impl for `str`, then I would expect that to break. Initially I was trying to figure out if such an impl would affect code one can write today, but I couldn't come up with anything.

Reply:
I would actually expect that both `str: Trait`  and `str: !Trait` in the same env doesn't cause issues as my view on negative trait bounds are just "a separate, but disjoint trait".

SO even with a `str: !Sized` impl you would still be able to prove `str: Sized` using the where-bound with my mental model. The current impl entirely ignores impls during assembly if they have a different polarity :thinking:

Reply:
> entirely ignores impls during assembly if they have a different polarity

oh right 🤦 yea these should never interact